// @storex/core/lib/core/sub-ins.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const subscriber_1 = require("./subscriber");
const dispatcher_1 = require("./dispatcher");
const subscribe_1 = require("./subscribe");
exports.des1 = new dispatcher_1.Dispatcher();
exports.des2 = new dispatcher_1.Dispatcher();
class DepIns extends subscriber_1.Subscriber {
    constructor() {
        super({ to: [{ dispatcher: exports.des1, name: "des1" }, exports.des2] });
        this.count1 = 0;
        this.count2 = 0;
    }
    update1() {
        this.count1++;
        console.log("update 1");
    }
    update2() {
        this.count2++;
        console.log("update 2");
    }
    updateAndDispatch2() {
        this.count2++;
        exports.des1.dispatch();
        console.log("update and dispatch 2");
    }
}
__decorate([
    subscribe_1.subscribe(["des1"])
], DepIns.prototype, "update1", null);
__decorate([
    subscribe_1.subscribe()
], DepIns.prototype, "update2", null);
__decorate([
    subscribe_1.subscribe()
], DepIns.prototype, "updateAndDispatch2", null);
exports.DepIns = DepIns;
