// @storex/core/lib/examples.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
var SortInsEvents;
(function (SortInsEvents) {
    SortInsEvents["AddItem"] = "add-item";
    SortInsEvents["DeleteItem"] = "delete-item";
})(SortInsEvents = exports.SortInsEvents || (exports.SortInsEvents = {}));
const events = Object.keys(SortInsEvents);
class StoreIns extends _1.Dispatcher {
    constructor() {
        super({ events });
        this.args1 = [];
        this.args1 = [];
    }
    action1(item) {
        this.args1.push(item);
        this.dispatch();
    }
    action2(items = []) {
        items.forEach(item => this.action1(item));
    }
    action3() { }
    action4() { }
}
__decorate([
    _1.dispatch()
], StoreIns.prototype, "args2", void 0);
__decorate([
    _1.dispatch([SortInsEvents.AddItem])
], StoreIns.prototype, "action1", null);
__decorate([
    _1.dispatch()
], StoreIns.prototype, "action2", null);
__decorate([
    _1.dispatch()
], StoreIns.prototype, "action3", null);
__decorate([
    _1.dispatch()
], StoreIns.prototype, "action4", null);
exports.StoreIns = StoreIns;
