// @storex/core/lib/main.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
var StoreInsEvents;
(function (StoreInsEvents) {
    StoreInsEvents["AddItem"] = "add-item";
    StoreInsEvents["DeleteItem"] = "delete-item";
})(StoreInsEvents = exports.StoreInsEvents || (exports.StoreInsEvents = {}));
const events = Object.keys(StoreInsEvents);
class StoreIns extends _1.Dispatcher {
    constructor() {
        super({ events });
        this.args1 = [];
        this.args1 = [];
    }
    action1(item) {
        this.args1.push(item);
        this.dispatch();
    }
    action2(items = []) {
        items.forEach(item => this.action1(item));
    }
}
__decorate([
    _1.dispatch()
], StoreIns.prototype, "args2", void 0);
__decorate([
    _1.dispatch([StoreInsEvents.AddItem])
], StoreIns.prototype, "action1", null);
__decorate([
    _1.dispatch()
], StoreIns.prototype, "action2", null);
exports.StoreIns = StoreIns;
const dispatcher = new StoreIns();
let count = 0;
const func = () => count++;
dispatcher.register(func);
dispatcher.dispatchOnce(() => {
    dispatcher.args2 = 4;
    dispatcher.args2 = 5;
    // dispatcher.action1("new item");
});
