// @storex/container/lib/filter/filter-function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
class FilterFunctions {
    constructor() {
        this.where = _1.where;
        this.sort = _1.sort;
        this.getPageItems = _1.getPageItems;
    }
}
exports.FilterFunctions = FilterFunctions;
function createDefaultItemsPerPage() {
    return function (items, filterData) {
        return _1.getPageItems(items, filterData.page, filterData.itemsPerPage);
    };
}
function createDefaultFreeSearch() {
    // var filterFunc ="" $filter("freeSearch");
    var filterFunc = freeSearch;
    return function (items, filterData, metaData) {
        // todo create general free search filter function
        var query = filterData.freeSearch;
        return filterFunc(items, query, metaData);
    };
}
// todo
function freeSearch(items, query, metaData) {
    if (!query)
        return items;
    query = query.toLowerCase();
    var newItems = items.filter(function (log) {
        // search..
    });
    return newItems;
}
